theano/gof/tests/test_types.py

```diff
@@ -76,3 +76,118 @@ def test_cdata():
+        self.op_chosen = op_to_const[choose_op]
+
+    def get_params(self, node):
+        return self.op_chosen
+
+    def make_node(self, a, b):
+        return Apply(self, [scalar.as_scalar(a), scalar.as_scalar(b)], [scalar.float64()])
+
+    def perform(self, node, inputs, outputs, op):
+        a, b = inputs
+        o, = outputs
+        if op == self.params_type.ADD:
+            o[0] = a + b
+        elif op == self.params_type.SUB:
+            o[0] = a - b
+        elif op == self.params_type.MULTIPLY:
+            o[0] = a * b
+        elif op == self.params_type.DIVIDE:
+            o[0] = a / b
+        else:
+            raise NotImplementedError('Unknown op id ' + str(op))
+
+    def c_code_cache_version(self):
+        return (1,)
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        a, b = inputs
+        o, = outputs
+        fail = sub['fail']
+        op = sub['params']
+        return """
+        switch((int)%(op)s) {
+            case ADD:
+                %(o)s = %(a)s + %(b)s;
+                break;
+            case SUB:
+                %(o)s = %(a)s - %(b)s;
+                break;
+            case MULTIPLY:
+                %(o)s = %(a)s * %(b)s;
+                break;
+            case DIVIDE:
+                %(o)s = %(a)s / %(b)s;
+                break;
+            default:
+                {%(fail)s}
+                break;
+        }
+        """ % locals()
+
+
+def test_enum():
+    # Check that invalid enum name raises exception.
+    for invalid_name in ('a', '_A', '0'):
+        try:
+            EnumList(invalid_name)
+        except AttributeError:
+            pass
+        else:
+            raise Exception('EnumList with invalid name should faild.')
+
+        try:
+            EnumType(**{invalid_name: 0})
+        except AttributeError:
+            pass
+        else:
+            raise Exception('EnumType with invalid name should fail.')
+
+    # Check that invalid enum value raises exception.
+    try:
+        EnumType(INVALID_VALUE='string is not allowe.')
+    except ValueError:
+        pass
+    else:
+        raise Exception('EnumType with invalid value should fail.')
+
+    # Check EnumType.
+    e1 = EnumType(C1=True, C2=12, C3=True, C4=-1, C5=False, C6=0.0)
+    e2 = EnumType(C1=1, C2=12, C3=1, C4=-1.0, C5=0.0, C6=0)
+    assert e1 == e2
+    assert not (e1 != e2)
+    assert hash(e1) == hash(e2)
+
+    # Test an op with EnumList.
+    a = scalar.int32()
+    b = scalar.int32()
+    c_add = TestOpEnumList('+')(a, b)
+    c_sub = TestOpEnumList('-')(a, b)
+    c_multiply = TestOpEnumList('*')(a, b)
+    c_divide = TestOpEnumList('/')(a, b)
+    f = theano.function([a, b], [c_add, c_sub, c_multiply, c_divide])
+    va = 12
+    vb = 15
+    ref_add = va + vb
+    ref_sub = va - vb
+    ref_multiply = va * vb
+    ref_divide = va // vb
+    ref = [ref_add, ref_sub, ref_multiply, ref_divide]
+    out = f(va, vb)
+    assert ref == out, (ref, out)
```

Comment:
PR for Wrapper is not yet merged, so Wrapper is not available in this branch.  I should wait for a PR to be merged, either this one or the PR for Wrapper, to make additional changes.


---

In any case, I will have to update Wrapper interface to handle constants, as the constant type is a bit special. Currently Wrapper interface is:
```python
wrapper = Wrapper(a=Scalar('int64'), 
                  b=Scalar('float32'),
                  c=Scalar('float64'))
```

I plan to extend this interface by allowing also the following one:
```python
wrapper = Wrapper(EnumList('CONSTANT_1', 'CONSTANT_2', 'CONSTANT_N'),
                  a=Scalar('int64'), 
                  b=Scalar('float32'),
                  c=Scalar('float64'))
```

ie., if the EnumType is declared in `*args` (not in `**kwargs`), then it will be treated as a special case by Wrapper, to that we could have:
```python
wrapper.CONSTANT_1
wrapper.CONSTANT_2
wrapper.CONSTANT_N
```
to make constants easy to access in Python.